orttraining/orttraining/python/training/torchdynamo/ort_backend.py

```diff
@@ -0,0 +1,519 @@
+    """
+    In torch.fx.Graph, placehoder is a special assignment node. If it's not
+    executed in the beginning, it could overwrite values computed by upstream
+    nodes.
+    """
+
+    graph = graph_module.graph
+    placeholders = []
+    first_not_placeholder = None
+    for node in graph.nodes:
+        if node.op == "placeholder":
+            placeholders.append(node)
+        if first_not_placeholder is None and node.op != "placeholder":
+            first_not_placeholder = node
+    if first_not_placeholder is None:
+        return
+    for placeholder in placeholders:
+        first_not_placeholder.prepend(placeholder)
+
+
+def _fx_to_torchscript(
+    fx_module: torch.fx.GraphModule,
+) -> torch.jit.ScriptModule:
+    for node in fx_module.graph.nodes:
+        if (
+            node.target == torch.ops.aten._to_copy  # type: ignore
+            and len(node.args) == 1
+            and len(node.kwargs) == 1
+            and "dtype" in node.kwargs
+        ):
+            node.target = torch.ops.aten.to  # type: ignore
+    for node in fx_module.graph.nodes:
+        new_kwargs = {}
+        for k, v in node.kwargs.items():
+            if isinstance(v, torch.device):
+                v = v.type
+            new_kwargs[k] = v
+        node.kwargs = new_kwargs
+    for node in fx_module.graph.nodes:
+        if isinstance(node.target, torch._ops.OpOverload):
+            node.target = node.target.overloadpacket
+    fx_module.graph.lint()
+    fx_module.recompile()
+    return torch.jit.script(fx_module)  # type: ignore
+
+
+def _decorate_script_module(script_module: torch.jit.ScriptModule, expected_inputs, expected_outputs):
+    for i, input_value in enumerate(script_module.graph.inputs()):  # type: ignore
+        if input_value.debugName() == "self":
+            script_module.graph.eraseInput(i)  # type: ignore
+            break
+    for i, input_value in enumerate(script_module.graph.inputs()):  # type: ignore
+        input_value.setType(torch._C.TensorType.create_from_tensor(expected_inputs[i]))
+    for i, output_value in enumerate(script_module.graph.outputs()):  # type: ignore
+        output_value.setType(torch._C.TensorType.create_from_tensor(expected_outputs[i]))
+
+
+def _create_onnx_proto(script_module):
+    onnx_proto = _jit_graph_to_onnx_model(script_module.graph, torch.onnx.OperatorExportTypes.ONNX)
+    return onnx_proto
+
+
+def _create_onnx_model(onnx_proto):
+    return onnx.ModelProto.FromString(onnx_proto)
+
+
+def _create_onnx_session(onnx_proto, device):
+    # TODO(wechi): Add more EPs per PyTorch device types.
+    # TODO(wechi): enable external allocators.
+    if device.type == "cuda":
+        return onnxruntime.InferenceSession(onnx_proto, providers=["CUDAExecutionProvider"])
+    return onnxruntime.InferenceSession(onnx_proto, providers=["CPUExecutionProvider"])
+
+
+def _get_onnx_devices(values: Tuple[torch.Tensor]) -> Tuple[ORTC.OrtDevice]:  # type: ignore
```

Comment:
`Tuple[torch.Tensor, ...]`

likely elsewhere too

Reply:
Curious why `type: ignore` is needed here?

Reply:
ORTC.OrtDevice is not detected.